File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.py
class Solution:
    def shortestPath(self, grid: List[List[int]], k: int) -> int:
        n, m = len(grid), len(grid[0])
        dirs = [
            (0,1),
            (1,0),
            (0,-1),
            (-1,0)
        ]

        if k >= n + m:
            return n + m - 2


        if grid[0][0] == 1 and k == 0:
            return -1


        vis = [[(k  + 1)for _ in range(m)] for _ in range(n)]
        need_remove_start = (grid[0][0] == 1)
        q = deque([(0, 0, need_remove_start, 0)])

        vis[0][0] = need_remove_start


        while q:
            row, col, removals, steps = q.popleft()

            if row == n-1 and col == m-1:
                return steps

            for dx, dy in dirs:
                next_row, next_col = row + dx, col + dy
                if next_row < n and next_col < m and next_row >= 0 and next_col >= 0:
                    need_remove_cell = (grid[next_row][next_col] == 1)
                    new_removals = removals + need_remove_cell
                    if vis[next_row][next_col] > new_removals:
                        vis[next_row][next_col] = new_removals
                        q.append((next_row, next_col, new_removals, steps + 1))


        return -1

        

File: 1293-shortest-path-in-a-grid-with-obstacles-elimination/1293-shortest-path-in-a-grid-with-obstacles-elimination.java
import java.util.*;

class Solution {
    public int shortestPath(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;

        // Optimization: If k allows us to take the direct Manhattan distance path
        if (k >= m + n - 2) {
            return m + n - 2;
        }

        // State: {row, col, remaining_k}
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0, k});

        // visited[row][col] stores the maximum remaining_k seen so far at that cell
        int[][] visited = new int[m][n];
        for (int[] row : visited) {
            Arrays.fill(row, -1);
        }
        visited[0][0] = k;

        int steps = 0;
        int[][] dirs = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

        while (!q.isEmpty()) {
            int size = q.size();
            
            for (int i = 0; i < size; i++) {
                int[] curr = q.poll();
                int r = curr[0];
                int c = curr[1];
                int rem = curr[2];

                // Target reached
                if (r == m - 1 && c == n - 1) {
                    return steps;
                }

                for (int[] dir : dirs) {
                    int nr = r + dir[0];
                    int nc = c + dir[1];

                    if (nr >= 0 && nr < m && nc >= 0 && nc < n) {
                        int newRem = rem - grid[nr][nc];

                        // Only proceed if we have removals left AND this path 
                        // provides more flexibility (higher k) than previous ones
                        if (newRem >= 0 && newRem > visited[nr][nc]) {
                            visited[nr][nc] = newRem;
                            q.offer(new int[]{nr, nc, newRem});
                        }
                    }
                }
            }
            steps++;
        }

        return -1;
    }
}